src/unix/uiactionx11.cpp

```diff
@@ -46,6 +55,113 @@ GtkWidget* wxGetTopLevelGTK();
 
 namespace
 {
+// Like the real events, this class tries to add a _fake_ delay to the generated
+// (fake) events so that chances that they got lost or ignored (for whatever reasons)
+// by the X server (or the WM, or even the input device driver) are minimized.
+class wxXSync
+{
+public:
+    wxXSync(wxX11Display& display)
+        : m_display(display), m_isMotion(true)
+    {
+    }
+
+    wxXSync(wxX11Display& display, bool depressed)
+        : m_display(display), m_isMotion(false)
+    {
+        static int s_maxDepressed = 0;
+
+        if ( depressed )
+        {
+            ++ms_numDepressed; // A key or button is down.
+
+            if ( ms_numDepressed - s_maxDepressed != 1 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
+
+            s_maxDepressed = ms_numDepressed;
+        }
+        else
+        {
+            if ( s_maxDepressed != 0 )
+            {
+                s_maxDepressed = 0; // reset
+
+                wxYield();
+                wxMilliSleep(Default_Delay);
+            }
+
+            --ms_numDepressed; // A key or button is up.
+
+            if ( ms_numDepressed < 0 )
+                wxFAIL_MSG("Invalid call to wxXSync() ctor");
+        }
+    }
+
+    ~wxXSync()
+    {
+        XSync(m_display, False);
+
+        m_isMotion ? OnMotion() : OnKeyOrButtonUp();
+    }
+
+private:
+    static void OnMotion()
+    {
+        wxMilliSleep(Motion_Delay);
+        wxYield();
+    }
+
+    static void OnKeyOrButtonUp()
+    {
+        if ( ms_numDepressed > 0 )
+        {
+            // Do nothing if a key / button is still depressed.
+            return;
+        }
+
+        wxYield();
+
+        // Default_Delay might not be sufficient for some
+        // tests which need more additionnal time to pass.
+        wxMilliSleep(Default_Delay + GetAdditionalDelay());
+    }
+
+    static long GetAdditionalDelay()
+    {
+        long delay;
+        wxString delayStr;
+
+        if ( wxGetEnv("WX_UI_TESTS_DELAY", &delayStr) &&
+             delayStr.ToCLong(&delay) )
+        {
+            delay = wxClip(delay, 0, 1000); // up to 1s
+        }
+        else
+        {
+            delay = 0;
+        }
+
+        return delay;
+    }
+
+private:
+    wxX11Display& m_display;
+    const bool    m_isMotion;
```

Comment:
I'd like to add a comment explaining that if it is `false`, it's button or key event because it isn't immediately obvious right now.